### ISSUE TITLE
seccomp: missing getrlimit from default whitelisting

### ISSUE THREAD

**Issue by lucab** [systemd/systemd#3970]
### Submission type
- Bug report
### systemd version the issue has been seen with

v229 (as part of rkt stage1-coreos)
### Used distribution

CoreOS Alpha 1097.0.0 (in rkt stage1)
### In case of bug report: Unexpected behaviour you saw

When running an app as a systemd service (inside rkt stage1) with strict seccomp whitelisting, spurious `getrlimit` calls appear (in straces) between seccomp filter setup and `execve`.

Original bugreport: https://github.com/coreos/rkt/issues/3074
### In case of bug report: Expected behaviour you didn't see

Default seccomp whitelisting should already include all syscalls needed for service setup.
### In case of bug report: Steps to reproduce the problem

Note: this may be reproducible with plain systemd units, but here below is the original report from rkt.

Running latest rkt (v1.12.0) with a stage1 using systemd and seccomp filtering as follow:

```
sudo strace -o rkt-seccomp-strace -ff rkt run --debug quay.io/coreos/alpine-sh --exec /bin/sleep -- 6
```

In the resulting strace you can see some additional `getrlimit` before `execve`. As they are past seccomp filter setup, users must be aware and manually whitelist those too:

```
prctl(PR_CAPBSET_DROP, CAP_AUDIT_READ)  = 0
capset({_LINUX_CAPABILITY_VERSION_3, 0}, {1<<CAP_CHOWN|1<<CAP_DAC_OVERRIDE|1<<CAP_DAC_READ_SEARCH|1<<CAP_FOWNER|1<<CAP_FSETID|1<<CAP_KILL|1<<CAP_SETGID|1<<CAP_SETUID|1<<CAP_SETPCAP|1<<CAP_LINUX_IMMUTABLE|1<<CAP_NET_BIND_SERVICE|1<<CAP_NET_BROADCAST|1<<CAP_NET_RAW|1<<CAP_IPC_OWNER|1<<CAP_SYS_CHROOT|1<<CAP_SYS_PTRACE|1<<CAP_SYS_ADMIN|1<<CAP_SYS_BOOT|1<<CAP_SYS_NICE|1<<CAP_SYS_RESOURCE|1<<CAP_SYS_TTY_CONFIG|1<<CAP_MKNOD|1<<CAP_LEASE|1<<CAP_AUDIT_WRITE|1<<CAP_AUDIT_CONTROL|1<<CAP_SETFCAP, 1<<CAP_CHOWN|1<<CAP_DAC_OVERRIDE|1<<CAP_DAC_READ_SEARCH|1<<CAP_FOWNER|1<<CAP_FSETID|1<<CAP_KILL|1<<CAP_SETGID|1<<CAP_SETUID|1<<CAP_SETPCAP|1<<CAP_LINUX_IMMUTABLE|1<<CAP_NET_BIND_SERVICE|1<<CAP_NET_BROADCAST|1<<CAP_NET_RAW|1<<CAP_IPC_OWNER|1<<CAP_SYS_CHROOT|1<<CAP_SYS_PTRACE|1<<CAP_SYS_ADMIN|1<<CAP_SYS_BOOT|1<<CAP_SYS_NICE|1<<CAP_SYS_RESOURCE|1<<CAP_SYS_TTY_CONFIG|1<<CAP_MKNOD|1<<CAP_LEASE|1<<CAP_AUDIT_WRITE|1<<CAP_AUDIT_CONTROL|1<<CAP_SETFCAP, 0}) = 0
setresuid(0, 0, 0)                      = 0
prctl(PR_GET_SECUREBITS)                = 0
brk(0x555a575c8000)                     = 0x555a575c8000
brk(0x555a575e9000)                     = 0x555a575e9000
brk(0x555a5760a000)                     = 0x555a5760a000
brk(0x555a5762b000)                     = 0x555a5762b000
brk(0x555a5764c000)                     = 0x555a5764c000
seccomp(SECCOMP_SET_MODE_STRICT, 1, NULL) = -1 EINVAL (Invalid argument)
seccomp(SECCOMP_SET_MODE_FILTER, 0, {len=816, filter=0x555a5762af90}) = 0
getrlimit(RLIMIT_STACK, {rlim_cur=8192*1024, rlim_max=RLIM64_INFINITY}) = 0
getrlimit(RLIMIT_STACK, {rlim_cur=8192*1024, rlim_max=RLIM64_INFINITY}) = 0
getrlimit(RLIMIT_STACK, {rlim_cur=8192*1024, rlim_max=RLIM64_INFINITY}) = 0
getrlimit(RLIMIT_STACK, {rlim_cur=8192*1024, rlim_max=RLIM64_INFINITY}) = 0
getrlimit(RLIMIT_STACK, {rlim_cur=8192*1024, rlim_max=RLIM64_INFINITY}) = 0
getrlimit(RLIMIT_STACK, {rlim_cur=8192*1024, rlim_max=RLIM64_INFINITY}) = 0
getrlimit(RLIMIT_STACK, {rlim_cur=8192*1024, rlim_max=RLIM64_INFINITY}) = 0
getrlimit(RLIMIT_STACK, {rlim_cur=8192*1024, rlim_max=RLIM64_INFINITY}) = 0
getrlimit(RLIMIT_STACK, {rlim_cur=8192*1024, rlim_max=RLIM64_INFINITY}) = 0
getrlimit(RLIMIT_STACK, {rlim_cur=8192*1024, rlim_max=RLIM64_INFINITY}) = 0
getrlimit(RLIMIT_STACK, {rlim_cur=8192*1024, rlim_max=RLIM64_INFINITY}) = 0
getrlimit(RLIMIT_STACK, {rlim_cur=8192*1024, rlim_max=RLIM64_INFINITY}) = 0
getrlimit(RLIMIT_STACK, {rlim_cur=8192*1024, rlim_max=RLIM64_INFINITY}) = 0
getrlimit(RLIMIT_STACK, {rlim_cur=8192*1024, rlim_max=RLIM64_INFINITY}) = 0
getrlimit(RLIMIT_STACK, {rlim_cur=8192*1024, rlim_max=RLIM64_INFINITY}) = 0
getrlimit(RLIMIT_STACK, {rlim_cur=8192*1024, rlim_max=RLIM64_INFINITY}) = 0
getrlimit(RLIMIT_STACK, {rlim_cur=8192*1024, rlim_max=RLIM64_INFINITY}) = 0
getrlimit(RLIMIT_STACK, {rlim_cur=8192*1024, rlim_max=RLIM64_INFINITY}) = 0
execve("/bin/sleep", ["/bin/sleep", "6"], [/* 6 vars */]) = 0
arch_prctl(ARCH_SET_FS, 0x7fda10216048) = 0
set_tid_address(0x7fda10216080)         = 5
mprotect(0x7fda10214000, 4096, PROT_READ) = 0
mprotect(0x55a365abb000, 16384, PROT_READ) = 0
getuid()                                = 0
nanosleep({6, 0}, 0x7fffcb878f78)       = 0
exit_group(0)  
```

**Comment by ronnychevalier** [comment]
By looking at the code of systemd (current and v229) and the code of coreos/systemd, neither call `getrlimit` after the `apply_seccomp` function. The only part of the code that could call this function with such parameters would be the code to apply the `LimitSTACK=xxxxx` by calling `setrlimit_closest` before setting the syscall whitelist.

So either I missed something, or you use a modified version of systemd, or the `execve` implementation used is calling `getrlimit` multiple times for an obscure reason?

**Comment by evverx** [comment]
This was fixed in 1280503b7e74

`strv_env_clean` -> `env_assignment_is_valid` -> `env_name_is_valid_n` ->

``` c
        if (n > ARG_MAX - 2)
                return false;
```

```
#ifndef ARG_MAX
#define ARG_MAX ((size_t) sysconf(_SC_ARG_MAX))
#endif
```

```
// glibc-2.22
// sysdeps/unix/sysv/linux/sysconf.c
...

    case _SC_ARG_MAX:
      /* Use getrlimit to get the stack limit.  */
      if (__getrlimit (RLIMIT_STACK, &rlimit) == 0)
        return MAX (legacy_ARG_MAX, rlimit.rlim_cur / 4);

      return legacy_ARG_MAX;
...
```

We have:

```
294   munmap(0x7f4e86b96000, 4096)      = 0
294   getrlimit(RLIMIT_STACK, {rlim_cur=8192*1024, rlim_max=RLIM64_INFINITY}) = 0
294   getrlimit(RLIMIT_STACK, {rlim_cur=8192*1024, rlim_max=RLIM64_INFINITY}) = 0
294   getrlimit(RLIMIT_STACK, {rlim_cur=8192*1024, rlim_max=RLIM64_INFINITY}) = 0
294   getrlimit(RLIMIT_STACK, {rlim_cur=8192*1024, rlim_max=RLIM64_INFINITY}) = 0
294   getrlimit(RLIMIT_STACK, {rlim_cur=8192*1024, rlim_max=RLIM64_INFINITY}) = 0
294   getrlimit(RLIMIT_STACK, {rlim_cur=8192*1024, rlim_max=RLIM64_INFINITY}) = 0
294   umask(022)                        = 0
294   chdir("/")                        = 0
294   stat("/proc/self/fd", {st_dev=makedev(0, 4), st_ino=10976, st_mode=S_IFDIR|0500, st_nlink=2, st_uid=0, st_gid=0, st_blksize=1024, st_blocks=0, st_size=0, st_atime=2016/08/$
4-23:24:01.651103494, st_mtime=2016/08/14-23:24:01.651103494, st_ctime=2016/08/14-23:24:01.651103494}) = 0
294   open("/proc/self/fd", O_RDONLY|O_NONBLOCK|O_DIRECTORY|O_CLOEXEC) = 3
294   fstat(3, {st_dev=makedev(0, 4), st_ino=10976, st_mode=S_IFDIR|0500, st_nlink=2, st_uid=0, st_gid=0, st_blksize=1024, st_blocks=0, st_size=0, st_atime=2016/08/14-23:24:01.65
1103494, st_mtime=2016/08/14-23:24:01.651103494, st_ctime=2016/08/14-23:24:01.651103494}) = 0
294   getdents(3, [{d_ino=10976, d_off=1, d_reclen=24, d_name=".", d_type=DT_DIR}, {d_ino=10975, d_off=2, d_reclen=24, d_name="..", d_type=DT_DIR}, {d_ino=10977, d_off=3, d_recle
n=24, d_name="0", d_type=DT_LNK}, {d_ino=10978, d_off=4, d_reclen=24, d_name="1", d_type=DT_LNK}, {d_ino=10979, d_off=5, d_reclen=24, d_name="2", d_type=DT_LNK}, {d_ino=10980, d_
off=66, d_reclen=24, d_name="3", d_type=DT_LNK}], 32768) = 144
294   getdents(3, [], 32768)            = 0
294   close(3)                          = 0
294   setrlimit(RLIMIT_NOFILE, {rlim_cur=1024, rlim_max=4*1024}) = 0
294   prctl(PR_GET_SECUREBITS)          = 0
294   capget({_LINUX_CAPABILITY_VERSION_3, 0}, NULL) = 0
294   capget({_LINUX_CAPABILITY_VERSION_3, 0}, {1<<CAP_CHOWN|1<<CAP_DAC_OVERRIDE|1<<CAP_DAC_READ_SEARCH|1<<CAP_FOWNER|1<<CAP_FSETID|1<<CAP_KILL|1<<CAP_SETGID|1<<CAP_SETUID|1<<CAP
_SETPCAP|1<<CAP_LINUX_IMMUTABLE|1<<CAP_NET_BIND_SERVICE|1<<CAP_NET_BROADCAST|1<<CAP_NET_ADMIN|1<<CAP_NET_RAW|1<<CAP_IPC_LOCK|1<<CAP_IPC_OWNER|1<<CAP_SYS_MODULE|1<<CAP_SYS_RAWIO|1
<<CAP_SYS_CHROOT|1<<CAP_SYS_PTRACE|1<<CAP_SYS_PACCT|1<<CAP_SYS_ADMIN|1<<CAP_SYS_BOOT|1<<CAP_SYS_NICE|1<<CAP_SYS_RESOURCE|1<<CAP_SYS_TIME|1<<CAP_SYS_TTY_CONFIG|1<<CAP_MKNOD|1<<CAP
_LEASE|1<<CAP_AUDIT_WRITE|1<<CAP_AUDIT_CONTROL|1<<CAP_SETFCAP|1<<CAP_MAC_OVERRIDE|1<<CAP_MAC_ADMIN|1<<CAP_SYSLOG|1<<CAP_WAKE_ALARM|1<<CAP_BLOCK_SUSPEND|1<<CAP_AUDIT_READ, 1<<CAP_
CHOWN|1<<CAP_DAC_OVERRIDE|1<<CAP_DAC_READ_SEARCH|1<<CAP_FOWNER|1<<CAP_FSETID|1<<CAP_KILL|1<<CAP_SETGID|1<<CAP_SETUID|1<<CAP_SETPCAP|1<<CAP_LINUX_IMMUTABLE|1<<CAP_NET_BIND_SERVICE
|1<<CAP_NET_BROADCAST|1<<CAP_NET_ADMIN|1<<CAP_NET_RAW|1<<CAP_IPC_LOCK|1<<CAP_IPC_OWNER|1<<CAP_SYS_MODULE|1<<CAP_SYS_RAWIO|1<<CAP_SYS_CHROOT|1<<CAP_SYS_PTRACE|1<<CAP_SYS_PACCT|1<<
CAP_SYS_ADMIN|1<<CAP_SYS_BOOT|1<<CAP_SYS_NICE|1<<CAP_SYS_RESOURCE|1<<CAP_SYS_TIME|1<<CAP_SYS_TTY_CONFIG|1<<CAP_MKNOD|1<<CAP_LEASE|1<<CAP_AUDIT_WRITE|1<<CAP_AUDIT_CONTROL|1<<CAP_S
ETFCAP|1<<CAP_MAC_OVERRIDE|1<<CAP_MAC_ADMIN|1<<CAP_SYSLOG|1<<CAP_WAKE_ALARM|1<<CAP_BLOCK_SUSPEND|1<<CAP_AUDIT_READ, 0}) = 0
294   seccomp(SECCOMP_SET_MODE_STRICT, 1, NULL <unfinished ...>
294   <... seccomp resumed> )           = -1 EINVAL (Invalid argument)
294   seccomp(SECCOMP_SET_MODE_FILTER, 0
```

**Comment by lucab** [comment]
@evverx thanks, I didn't manage to uncover that codepath while debugging today, but the `sysconf()` side-effect makes sense. I'll re-check on rkt side as soon as we bump systemd in stage1, but I'm quite confident this nailed it.

**Comment by evverx** [comment]
Well, I've found another [issue](https://github.com/systemd/systemd/issues/3962#issuecomment-239722151):

> audit: type=1326 audit(1471109203.980:49): auid=4294967295 uid=0 gid=0 ses=4294967295 pid=280 comm="(sh)" exe="/usr/lib/systemd/systemd" sig=31 arch=c000003e syscall=41 compat=0 ip=0x7fbe90427757 code=0x0

We call `log_open` [there](https://github.com/systemd/systemd/blob/d87a2ef78277177b363f3d603221183d22811a82/src/core/execute.c#L2420).

``` sh
$ sudo stap -v  --ldd -e 'probe process("/usr/lib/systemd/systemd").syscall { if ($syscall != 41) next; printf("%s\n%s\n--\n\n", thread_indent(0), sprint_ubacktrace()) }'

--                                                                                                                                                                       
     0 (sh)(2744):
socket+0x7 [libc-2.22.so]
create_log_socket+0x29 [systemd]
log_open+0xda [systemd]
exec_child+0x2ae6 [systemd]
exec_spawn+0x736 [systemd]
service_spawn+0x37e [systemd]
service_enter_start+0xd8 [systemd]
service_start.lto_priv.347+0x30c [systemd]
manager_dispatch_run_queue.lto_priv.931+0xa1c [systemd]
source_dispatch.lto_priv.917+0x12d [systemd]
manager_loop+0x6b7 [systemd]
main+0x1bab [systemd]
__libc_start_main+0xf0 [libc-2.22.so]
_start+0x29 [systemd]
--
```

Seems like we should remove/reorder that `log_open` (and [`setexeccon`](https://github.com/systemd/systemd/blob/d87a2ef78277177b363f3d603221183d22811a82/src/core/execute.c#L2389), and [`aa_change_onexec`](https://github.com/systemd/systemd/blob/d87a2ef78277177b363f3d603221183d22811a82/src/core/execute.c#L2400))

@ronnychevalier , @topimiettinen , what do you think?

BTW:
@topimiettinen , I've just seen https://lists.freedesktop.org/archives/systemd-devel/2016-August/037310.html
Can you push that script somewhere?  github.com/topimiettinen/systemd-settings-generator or so.
I'll take a look. Thanks!

**Comment by poettering** [comment]
given that getrlimit() is frequently used to determine stack size and stuff like that I figure getrlimit() really should be among the implicitly OK syscalls.

**Comment by topimiettinen** [comment]
Ideally there should be no system calls between the final seccomp() and execve() to minimize restrictions and the default set. I'd suppose SELinux and AppArmor also would like to be the last ones too, so there's conflict. It's also not a good idea to block system calls that setexeccon() and aa_change_onexec happen to use. This could be avoided if the kernel provided a new way to install seccomp filters so that they only take effect after execve(), much like the context set by setexeccon() applies after execve(). Then also execve() could be allowed to be blocked by a filter to prevent the service from executing anything else, which would be very nice.

**Comment by topimiettinen** [comment]
@evverx : I created https://github.com/topimiettinen/systemd-settings-generator.

**Comment by poettering** [comment]
I added a commit to #3984 which adds getrlimit() to the syscalls whitelisted by default, which should fix the issue at hand properly. I think it's completely safe to query one's own resource limits, hence this really belongs in the whitelist.

@evverx can you open a new bug for the post-seccomp logging fuckup when we try to issue socket() after having seccomp installed? My guess is that we should first setup selinux/aa, and only then do seccomp, and really make sure that if seccomp is set up we don't do any further syscalls besides exit(). This of course means we also won't do proper logging in that case anymore, but that's an OK deal and not a complete loss as seccomp failures get their own exit code with "systemctl status" shows nicely anyway...

**Comment by poettering** [comment]
Ah, well, I figured I could create an issue for that story myself, in #3993. Let's follow-up with the discussion there.
